File: prj6/MaxHeap.py
class MaxHeap:
    def __init__(self):
        self.heap_array = list()
        self.order = 0

    def insert(self, data):
        self.heap_array.append(data)
        self.order += 1
        self.heap_insert_order(len(self.heap_array) - 1)

    def heap_insert_order(self, order):
        parent = int((order - 1) / 2)

        if self.heap_array[parent] > 0:
            if self.heap_array[order] > self.heap_array[parent]:
                self.heap_array[order], self.heap_array[parent] = self.heap_array[parent], self.heap_array[order]

                self.heap_insert_order(parent)

    def remove(self, data):
        pass

    def heap_remove_order(self, order, root_index):
        left_child = (2 * order) + 1
        right_child = (2 * order) + 2

        root = root_index

        pass

    def display(self):
        print(self.heap_array)


sample = MaxHeap()


sample.insert(40)
sample.insert(30)
sample.insert(24)
sample.insert(10)
sample.insert(20)
sample.insert(5)
sample.insert(3)


sample.display()

File: prj6/maxheap.cpp
#include <iostream>

using namespace std;
class MaxHeap {
private:
    int* heapArray;
    int capacity;
    int size;

public:
    MaxHeap(int cap) : capacity(cap), size(0) {
        heapArray = new int[capacity];
    }

    ~MaxHeap() {
        delete[] heapArray;
    }

    void insert(int value) {
        if (size == capacity) {
            cout << "Heap overflow\n";
            return;
        }

        size++;
        int i = size - 1;
        heapArray[i] = value;


        while (i > 0 && heapArray[parent(i)] < heapArray[i]) {
            swap(heapArray[i], heapArray[parent(i)]);
            i = parent(i);
        }
    }

    int extractMax() {
        if (size <= 0) {
            cout << "Heap underflow\n";
    
        }

        if (size == 1) {
            size--;
            return heapArray[0];
        }

        int root = heapArray[0];
        heapArray[0] = heapArray[size - 1];
        size--;


        heapifyDown(0);

        return root;
    }

    void printHeap() {
        for (int i = 0; i < size; i++) {
            cout << heapArray[i] << " ";
        }
        cout << "\n";
    }

private:
    int parent(int i) {
        return (i - 1) / 2;
    }

    void swap(int &a, int &b) {
        int temp = a;
        a = b;
        b = temp;
    }

    void heapifyDown(int i) {
        int maxIndex = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < size && heapArray[left] > heapArray[maxIndex]) {
            maxIndex = left;
        }

        if (right < size && heapArray[right] > heapArray[maxIndex]) {
            maxIndex = right;
        }

        if (i != maxIndex) {
            swap(heapArray[i], heapArray[maxIndex]);
            heapifyDown(maxIndex);
        }
    }
};
